README.rst
===============================
Realtime Stock
===============================


.. image:: https://badge.fury.io/py/realtime-stock.svg
        :target: https://pypi.python.org/pypi/realtime-stock

.. image:: https://img.shields.io/travis/condereis/realtime-stock.svg
        :target: https://travis-ci.org/condereis/realtime-stock

.. image:: https://readthedocs.org/projects/realtime-stock/badge/?version=latest
        :target: https://realtime-stock.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/condereis/realtime-stock/shield.svg
     :target: https://pyup.io/repos/github/condereis/realtime-stock/
     :alt: Updates


Realtime Stock is a Python package to gather realtime stock quotes from Yahoo Finance. The package enables you to handle single stocks or portfolios, optimizing the nunber of requests necessary to gather quotes for a large number of stocks.


* Documentation: https://realtime-stock.readthedocs.io.
* Free software: MIT license


Features
--------

**Stock class**

* Method to get stock's latest price.
* Method to get all stock's information provided by Yahoo Finance.
* Method get stock's daily historical information.
* Method download stock's historical data from Yahoo Finance.

**Utility functions**

* Function to request recent quotes about a list of tickers.
* Function to get stock's daily historical information.
* Function to download historical data about a list of tickers.


Installation
------------
To install Realtime Stock, run this command in your terminal:

.. code:: bash

    $ pip install realtime-stock

Check `here <https://realtime-stock.readthedocs.io/en/latest/installation.html>`_  for further information on installation.


Basic Usage
-----------

For single stocks:

.. code:: python

	>>> from rtstock.stock import Stock
	>>> stock = Stock('AAPL')
	>>> stock.get_latest_price()
	{
	    'LastTradePriceOnly': '95.89',
	    'LastTradeTime': '4:00pm'
	}

Check `here <https://realtime-stock.readthedocs.io/en/latest/usage.html>`_  for further information on usage.

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage



HISTORY.rst
=======
History
=======

1.0.0 (2016-07-27)
------------------

* First release on PyPI.
* Stock class implemented.
* Function to request recent quotes about a list of tickers.
* Function to get stock's daily historical information.
* Function to download historical data about a list of tickers..

tests/__init__.py
# -*- coding: utf-8 -*-

import sys
if sys.version_info < (3,0):
	# Python 2
	import test_stock

tests/test_stock.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_stock
----------------------------------

Tests for `rtstock` module.
"""


import os
import sys
import unittest

import rtstock.error as error
from rtstock.stock import Stock


class TestRealStock(unittest.TestCase):
    """Tests for Stock."""

    def setUp(self):
        """SetUp."""
        self.ticker = 'AAPL'
        self.stock = Stock(self.ticker)

    def test_get_ticker(self):
        """Test get_ticker."""
        self.assertEqual(self.stock.get_ticker(), self.ticker)

    def test_set_ticker(self):
        """Test set_ticker."""
        self.stock.set_ticker('new_ticker')
        self.assertEqual(self.stock.get_ticker(), 'new_ticker')

    def test_get_latest_price(self):
        """Test get_latest_price success."""
        response = self.stock.get_latest_price()
        # Checks for 'LastTradePriceOnly', 'LastTradeTime' field
        self.assertEqual(len(response[0].keys()), 2)
        for key in response[0].keys():
            self.assertTrue(key in ['LastTradePriceOnly', 'LastTradeTime'])

    def test_get_info(self):
        """Test get_info success."""
        keys = ['Ask', 'AverageDailyVolume', 'Bid', 'BookValue', 'Change',
                'Change_PercentChange', 'ChangeFromFiftydayMovingAverage',
                'ChangeFromTwoHundreddayMovingAverage',
                'ChangeFromYearHigh', 'ChangeFromYearLow',
                'ChangeinPercent', 'Currency', 'DaysHigh', 'DaysLow',
                'DaysRange', 'DividendPayDate', 'DividendShare',
                'DividendYield', 'EarningsShare', 'EBITDA',
                'EPSEstimateCurrentYear', 'EPSEstimateNextQuarter',
                'EPSEstimateNextYear', 'ExDividendDate',
                'FiftydayMovingAverage', 'LastTradeDate',
                'LastTradePriceOnly', 'LastTradeTime', 'LastTradeWithTime',
                'MarketCapitalization', 'Name', 'OneyrTargetPrice', 'Open',
                'PEGRatio', 'PERatio', 'PercebtChangeFromYearHigh',
                'PercentChange', 'PercentChangeFromFiftydayMovingAverage',
                'PercentChangeFromTwoHundreddayMovingAverage',
                'PercentChangeFromYearLow', 'PreviousClose', 'PriceBook',
                'PriceEPSEstimateCurrentYear', 'PriceEPSEstimateNextYear',
                'PriceSales', 'ShortRatio', 'StockExchange', 'Symbol',
                'TwoHundreddayMovingAverage', 'Volume', 'YearHigh',
                'YearLow', 'YearRange']

        response = self.stock.get_info()
        # Checks for 'LastTradePriceOnly', 'LastTradeTime' field
        self.assertEqual(len(response[0].keys()), len(keys))
        for key in response[0].keys():
            self.assertTrue(key in keys)

    def test_save_historical(self):
        """Test save_historical success."""
        first_line = b'Date,Open,High,Low,Close,Volume,Adj Close\n'
        self.stock.save_historical('.')
        with open(self.ticker + '.csv', 'rb') as f:
            self.assertEqual(f.readline(), first_line)
        os.remove(self.ticker + '.csv')


class TestFakeStock(unittest.TestCase):
    """Tests for fake Stock"""

    def setUp(self):
        """SetUp."""
        self.ticker = 'fake_ticker'
        self.stock = Stock(self.ticker)

    def test_get_ticker(self):
        """Test get_ticker."""
        self.assertEqual(self.stock.get_ticker(), self.ticker)

    def test_set_ticker(self):
        """Test set_ticker."""
        self.stock.set_ticker('new_ticker')
        self.assertEqual(self.stock.get_ticker(), 'new_ticker')

    def test_get_latest_price(self):
        """Test get_latest_price error."""
        response = self.stock.get_latest_price()
        # Checks for 'LastTradePriceOnly', 'LastTradeTime' field
        # Check if both fields are empty
        self.assertEqual(len(response[0].keys()), 2)
        for key in response[0].keys():
            self.assertTrue(key in ['LastTradePriceOnly', 'LastTradeTime'])
            self.assertFalse(response[0][key])

    def test_get_info(self):
        """Test get_info error."""
        keys = ['Ask', 'AverageDailyVolume', 'Bid', 'BookValue', 'Change',
                'Change_PercentChange', 'ChangeFromFiftydayMovingAverage',
                'ChangeFromTwoHundreddayMovingAverage',
                'ChangeFromYearHigh', 'ChangeFromYearLow',
                'ChangeinPercent', 'Currency', 'DaysHigh', 'DaysLow',
                'DaysRange', 'DividendPayDate', 'DividendShare',
                'DividendYield', 'EarningsShare', 'EBITDA',
                'EPSEstimateCurrentYear', 'EPSEstimateNextQuarter',
                'EPSEstimateNextYear', 'ExDividendDate',
                'FiftydayMovingAverage', 'LastTradeDate',
                'LastTradePriceOnly', 'LastTradeTime', 'LastTradeWithTime',
                'MarketCapitalization', 'Name', 'OneyrTargetPrice', 'Open',
                'PEGRatio', 'PERatio', 'PercebtChangeFromYearHigh',
                'PercentChange', 'PercentChangeFromFiftydayMovingAverage',
                'PercentChangeFromTwoHundreddayMovingAverage',
                'PercentChangeFromYearLow', 'PreviousClose', 'PriceBook',
                'PriceEPSEstimateCurrentYear', 'PriceEPSEstimateNextYear',
                'PriceSales', 'ShortRatio', 'StockExchange', 'Symbol',
                'TwoHundreddayMovingAverage', 'Volume', 'YearHigh',
                'YearLow', 'YearRange']

        response = self.stock.get_info()
        # Checks for 'LastTradePriceOnly', 'LastTradeTime' field
        self.assertEqual(len(response[0].keys()), len(keys))
        for key in response[0].keys():
            self.assertTrue(key in keys)
            # Check if all fields but Symbol are empty
            if key == 'Symbol':
                self.assertEqual(response[0][key], self.ticker)
            else:
                self.assertFalse(response[0][key])

    def test_save_historical(self):
        """Test save_historical error."""
        with self.assertRaises(error.RequestError):
            self.stock.save_historical('.')


class TestStockGetHistorical(unittest.TestCase):
    """Tests for Stock's get_historical method."""

    def setUp(self):
        """SetUp."""
        self.ticker = 'AAPL'
        self.stock = Stock(self.ticker)
        self.start_date = '2016-03-01'
        self.end_date = '2016-03-03'
        self.keys = ['Date', 'Adj_Close', 'Close',
                     'High', 'Low', 'Open', 'Volume']

    def test_success(self):
        """Test get_historical success."""
        response = self.stock.get_historical(self.start_date, self.end_date)
        # Checks for 3 dictionaries (one per day)
        self.assertEqual(len(response), 3)
        # Checks for the field on 'keys'
        self.assertEqual(response[0]['Date'], self.end_date)
        self.assertEqual(response[2]['Date'], self.start_date)
        self.assertEqual(len(response[0].keys()), len(self.keys))
        for key in response[0].keys():
            self.assertTrue(key in self.keys)

    def test_wrong_format(self):
        """Test get_historical passing a wrong date."""
        with self.assertRaises(ValueError):
            self.stock.get_historical(self.start_date, '11111')

    def test_more_than_a_year(self):
        """Test get_historical passing a period greater than a year."""
        with self.assertRaises(ValueError):
            self.stock.get_historical(self.start_date, '2017-03-03')

    def test_change_start_and_end(self):
        """Test get_historical misplacing start and end dates."""
        with self.assertRaises(ValueError):
            self.stock.get_historical(self.end_date, self.start_date)


if __name__ == '__main__':
    sys.exit(unittest.main())


tests/test_utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_utils
----------------------------------

Tests for `utils` module.
"""

import os
import sys
import unittest

import rtstock.error as error
import rtstock.utils as utils


class TestRequestQuotes(unittest.TestCase):
    """Tests for request_quotes function."""

    def setUp(self):
        """SetUp."""
        self.tickers_list = ['AAPL', 'YHOO']
        self.fields = ['Name', 'PreviousClose']

    def test_success(self):
        """Test request_quotes success."""
        response = utils.request_quotes(self.tickers_list, self.fields)
        # Checks for 2 dictionaries
        self.assertEqual(len(response), len(self.tickers_list))
        # Checks for 'Name', 'PreviousClose' field
        self.assertEqual(len(response[0].keys()), len(self.fields))
        for key in response[0].keys():
            self.assertTrue(key in self.fields)

    def test_fake_column(self):
        """Test request_quotes asking for a fake column."""
        # Invalid among valid fields
        fields = ['Name', 'invalid_field']
        response = utils.request_quotes(self.tickers_list, fields)
        self.assertEqual(len(response[0].keys()), 1)
        # Only invalid fields
        fields = ['invalid_field_1', 'invalid_field_2']
        with self.assertRaises(error.RequestError):
            response = utils.request_quotes(self.tickers_list, fields)

    def test_fake_company(self):
        """Test request_quotes asking for an invalid company."""
        response = utils.request_quotes(['fake_company'], self.fields)
        # Checks for 2 dictionaries
        self.assertEqual(len(response), 1)
        # Checks for 'Name', 'PreviousClose' field
        self.assertEqual(len(response[0].keys()), len(self.fields))
        for key in response[0].keys():
            self.assertTrue(key in self.fields)
        # Check if both fields are empty
        self.assertFalse(response[0]['Name'])
        self.assertFalse(response[0]['PreviousClose'])


class TestRequestHistorical(unittest.TestCase):
    """Tests for request_historical function."""

    def setUp(self):
        """SetUp."""
        self.tickers_list = ['AAPL', 'YHOO']
        self.start_date = '2016-03-01'
        self.end_date = '2016-03-03'
        self.keys = ['Date', 'Adj_Close', 'Close',
                     'High', 'Low', 'Open', 'Volume']

    def test_success(self):
        """Test request_historical success."""
        response = utils.request_historical(
            self.tickers_list[0], self.start_date, self.end_date)
        # Checks for 3 dictionaries (one per day)
        self.assertEqual(len(response), 3)
        # Checks for the field on 'keys'
        self.assertEqual(response[0]['Date'], self.end_date)
        self.assertEqual(response[2]['Date'], self.start_date)
        self.assertEqual(len(response[0].keys()), len(self.keys))
        for key in response[0].keys():
            self.assertTrue(key in self.keys)

    def test_wrong_format(self):
        """Test request_historical passing a wrong date."""
        with self.assertRaises(ValueError):
            utils.request_historical(self.tickers_list[0],
                                     self.start_date, '11111')

    def test_more_than_a_year(self):
        """Test request_historical passing a period greater than a year."""
        with self.assertRaises(ValueError):
            utils.request_historical(self.tickers_list[0],
                                     self.start_date, '2017-03-03')

    def test_change_start_and_end(self):
        """Test request_historical misplacing start and end dates."""
        with self.assertRaises(ValueError):
            utils.request_historical(self.tickers_list[0],
                                     self.end_date, self.start_date)


class TestDownloadHistorical(unittest.TestCase):
    """Tests for download_historical function."""

    def setUp(self):
        """SetUp."""
        self.tickers_list = ['AAPL', 'YHOO']
        os.makedirs('test_files')
        self.output_folder = 'test_files'

    def test_success(self):
        """Test download_historical success."""
        first_line = b'Date,Open,High,Low,Close,Volume,Adj Close\n'
        utils.download_historical(self.tickers_list, self.output_folder)
        with open(os.path.join(self.output_folder,
                               self.tickers_list[0] + '.csv'), 'rb') as f:
            self.assertEqual(f.readline(), first_line)

    def test_invalid_company(self):
        """Test download_historical with invalid company."""
        with self.assertRaises(error.RequestError):
            utils.download_historical(['fake_company'], self.output_folder)

    def tearDown(self):
        """Cleaning up."""
        for root, dirs, files in os.walk(self.output_folder, topdown=False):
            for name in files:
                os.remove(os.path.join(root, name))
            for name in dirs:
                os.rmdir(os.path.join(root, name))
        os.rmdir(self.output_folder)

if __name__ == '__main__':
    sys.exit(unittest.main())


rtstock/utils.py
"""
Utility functions.

This module contains utility functions to gather information
from Yahoo Finance.
"""

from __future__ import unicode_literals
import datetime
import json
import os

try:
    # Python 3
    from urllib.request import urlopen, quote, urlretrieve
except ImportError:
    # Python 2
    from urllib2 import urlopen, quote
    from urllib import urlretrieve

from .error import RequestError


def __validate_list(list_to_validate):
    """Validate list."""
    if not type(list_to_validate) is list:
        raise TypeError(
            "List expected, " +
            type(list_to_validate).__name__ + " found."
        )


def __validate_dates(start_date, end_date):
    """Validate if a date string.

    Validate if a string is a date on yyyy-mm-dd format and it the
    period between them is less than a year.
    """
    try:
        start_date = datetime.datetime.strptime(start_date, '%Y-%m-%d')
        end_date = datetime.datetime.strptime(end_date, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect data format, should be yyyy-mm-dd")
    if (end_date - start_date).days > 366:
        raise ValueError("The difference between start and end date " +
                         "should be less than or equal to 366 days.")
    if (end_date - start_date).days < 0:
        raise ValueError("End date cannot be before start date.")


def __yahoo_request(query):
    """Request Yahoo Finance information.

    Request information from YQL.
    `Check <http://goo.gl/8AROUD>`_ for more information on YQL.
    """
    query = quote(query)
    url = 'https://query.yahooapis.com/v1/public/yql?q=' + query + \
        '&format=json&env=store://datatables.org/alltableswithkeys'

    response = urlopen(url).read()

    return json.loads(response.decode('utf-8'))['query']['results']


def request_quotes(tickers_list, selected_columns=['*']):
    """Request Yahoo Finance recent quotes.

    Returns quotes information from YQL. The columns to be requested are
    listed at selected_columns. Check `here <http://goo.gl/8AROUD>`_ for more
    information on YQL.

    >>> request_quotes(['AAPL'], ['Name', 'PreviousClose'])
    {
        'PreviousClose': '95.60',
        'Name': 'Apple Inc.'
    }

    :param table: Table name.
    :type table: string
    :param tickers_list: List of tickers that will be returned.
    :type tickers_list: list of strings
    :param selected_columns: List of columns to be returned, defaults to ['*']
    :type selected_columns: list of strings, optional
    :returns: Requested quotes.
    :rtype: json
    :raises: TypeError, TypeError
    """
    __validate_list(tickers_list)
    __validate_list(selected_columns)
    query = 'select {cols} from yahoo.finance.quotes where symbol in ({vals})'
    query = query.format(
        cols=', '.join(selected_columns),
        vals=', '.join('"{0}"'.format(s) for s in tickers_list)
    )

    response = __yahoo_request(query)

    if not response:
        raise RequestError('Unable to process the request. Check if the ' +
                           'columns selected are valid.')

    if not type(response['quote']) is list:
        return [response['quote']]
    return response['quote']


def request_historical(ticker, start_date, end_date):
    """Get stock's daily historical information.

    Returns a dictionary with Adj Close, Close, High, Low, Open and
    Volume, between the start_date and the end_date. Is start_date and
    end_date were not provided all the available information will be
    retrieved. Information provided by YQL platform.
    Check `here <http://goo.gl/8AROUD>`_ for more information on YQL.

    .. warning:: Request limited to a period not greater than 366 days.
        Use download_historical() to download the full historical data.


    >>> request_historical('AAPL', '2016-03-01', '2016-03-02')
    [
        {
            'Close': '100.75',
            'Low': '99.639999',
            'High': '100.889999',
            'Adj_Close': '100.140301',
            'Date': '2016-03-02',
            'Open': '100.510002',
            'Volume': '33169600'
        },
        {
            'Close': '100.529999',
            'Low': '97.419998',
            'High': '100.769997',
            'Adj_Close': '99.921631',
            'Date': '2016-03-01',
            'Open': '97.650002',
            'Volume': '50407100'
        }
    ]

    :param start_date: Start date
    :type start_date: string on the format of "yyyy-mm-dd"
    :param end_date: End date
    :type end_date: string on the format of "yyyy-mm-dd"
    :returns: Daily historical information.
    :rtype: list of dictionaries
    """
    __validate_dates(start_date, end_date)

    cols = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj_Close']
    query = 'select {cols} from yahoo.finance.historicaldata ' + \
        'where symbol in ("{ticker}") and startDate = "{start_date}" ' + \
        'and endDate = "{end_date}"'
    query = query.format(
        cols=', '.join(cols),
        ticker=ticker,
        start_date=start_date,
        end_date=end_date
    )

    response = __yahoo_request(query)
    if not response:
        raise RequestError('Unable to process the request. Check if the ' +
                           'stock ticker used is a valid one.')

    if not type(response['quote']) is list:
        return [response['quote']]
    return response['quote']


def download_historical(tickers_list, output_folder):
    """Download historical data from Yahoo Finance.

    Downloads full historical data from Yahoo Finance as CSV. The following
    fields are available: Adj Close, Close, High, Low, Open and Volume. Files
    will be saved to output_folder as <ticker>.csv.

    :param tickers_list: List of tickers that will be returned.
    :type tickers_list: list of strings
    :param output_folder: Output folder path
    :type output_folder: string
    """
    __validate_list(tickers_list)
    for ticker in tickers_list:
        file_name = os.path.join(output_folder, ticker + '.csv')
        with open(file_name, 'wb') as f:
            base_url = 'http://real-chart.finance.yahoo.com/table.csv?s='
            try:
                urlopen(base_url + ticker)
                urlretrieve(base_url + ticker, f.name)
            except:
                os.remove(file_name)
                raise RequestError('Unable to process the request. Check if ' +
                                   ticker + ' is a valid stock ticker')



travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'condereis/realtime-stock'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


rtstock/stock.py
"""
Stock module.

This module contains all the classes used to retrieve information
about a single stock from Yahoo Finances. That includes, real-time
quotes information as well as historical data.
"""

from __future__ import unicode_literals
from .utils import request_quotes, request_historical, download_historical
from .error import RequestError


class Stock(object):
    """Class for handling stock.

    Provides methods to retrieve real-time quotes and historical
    data from Yahoo Finance database.

    >>> from rtstock.stock
-  import Stock
    >>>
    >>> stock = Stock('AAPL')
    >>> print(stock)
    <Stock AAPL>

    :param ticker: Stock ticker in Yahoo Finances format.
    :type ticker: string
    """

    def __init__(self, ticker):
        """Instantiate Stock class."""
        self.__ticker = ticker

    def __repr__(self):
        """An unambiguous representation of a Stock's instance."""
        return '<Stock {ticker}>'.format(ticker=self.__ticker)

    def __eq__(self, other):
        """Equality comparison operator."""
        if isinstance(other, Stock):
            return self.__repr__() == other.__repr__()
        return False

    def __ne__(self, other):
        """Inquality comparison operator."""
        return not self.__eq__() == other.__repr__()

    def __hash__(self):
        """Hash representation of a Stock's instance."""
        return hash(self.__repr__())

    def get_ticker(self):
        """Get stock's ticker.

        >>> stock.get_ticker()
        'AAPL'

        :returns: Ticker.
        :rtype: string
        """
        return self.__ticker

    def set_ticker(self, ticker):
        """Set stock's ticker.

        >>> stock.set_ticker('YHOO')
        >>> print(stock)
        <Stock YHOO>

        :param ticker: Stock ticker in Yahoo Finances format.
        :type ticker: string
        """
        self.__ticker = ticker

    def get_latest_price(self):
        """Get stock's latest price.

        Get the latest available quote from Yahoo Finance along with its
        respective time.

        >>> stock.get_latest_price()
        {
            'LastTradePriceOnly': '95.89',
            'LastTradeTime': '4:00pm'
        }

        :returns: Dictionary with latest price and trade time.
        :rtype: dictionary
        """
        return request_quotes([self.__ticker],
                              ['LastTradePriceOnly', 'LastTradeTime'])

    def get_info(self):
        """Get all stock's information provided by Yahoo Finance.

        There is no guarantee that all the fields will be available for all
        stocks. That being said, the following fields will be retrieved by
        this method as a python dictionary from YQL platform:

        - Ask
        - AverageDailyVolume
        - Bid
        - BookValue
        - Change
        - Change_PercentChange
        - ChangeFromFiftydayMovingAverage
        - ChangeFromTwoHundreddayMovingAverage
        - ChangeFromYearHigh
        - ChangeFromYearLow
        - ChangeinPercent
        - Currency
        - DaysHigh
        - DaysLow
        - DaysRange
        - DividendPayDate
        - DividendShare
        - DividendYield
        - EarningsShare
        - EBITDA
        - EPSEstimateCurrentYear
        - EPSEstimateNextQuarter
        - EPSEstimateNextYear
        - ExDividendDate
        - FiftydayMovingAverage
        - LastTradeDate
        - LastTradePriceOnly
        - LastTradeTime
        - LastTradeWithTime
        - MarketCapitalization
        - Name
        - OneyrTargetPrice
        - Open
        - PEGRatio
        - PERatio
        - PercebtChangeFromYearHigh
        - PercentChange
        - PercentChangeFromFiftydayMovingAverage
        - PercentChangeFromTwoHundreddayMovingAverage
        - PercentChangeFromYearLow
        - PreviousClose
        - PriceBook
        - PriceEPSEstimateCurrentYear
        - PriceEPSEstimateNextYear
        - PriceSales
        - ShortRatio
        - StockExchange
        - Symbol
        - TwoHundreddayMovingAverage
        - Volume
        - YearHigh
        - YearLow
        - YearRange

        Check `here <http://goo.gl/8AROUD>`_ for more information on YQL.

        :returns: Dictionary with all the available information.
        :rtype: dictionary
        """
        keys = ['Ask', 'AverageDailyVolume', 'Bid', 'BookValue', 'Change',
                'Change_PercentChange', 'ChangeFromFiftydayMovingAverage',
                'ChangeFromTwoHundreddayMovingAverage',
                'ChangeFromYearHigh', 'ChangeFromYearLow',
                'ChangeinPercent', 'Currency', 'DaysHigh', 'DaysLow',
                'DaysRange', 'DividendPayDate', 'DividendShare',
                'DividendYield', 'EarningsShare', 'EBITDA',
                'EPSEstimateCurrentYear', 'EPSEstimateNextQuarter',
                'EPSEstimateNextYear', 'ExDividendDate',
                'FiftydayMovingAverage', 'LastTradeDate',
                'LastTradePriceOnly', 'LastTradeTime', 'LastTradeWithTime',
                'MarketCapitalization', 'Name', 'OneyrTargetPrice', 'Open',
                'PEGRatio', 'PERatio', 'PercebtChangeFromYearHigh',
                'PercentChange', 'PercentChangeFromFiftydayMovingAverage',
                'PercentChangeFromTwoHundreddayMovingAverage',
                'PercentChangeFromYearLow', 'PreviousClose', 'PriceBook',
                'PriceEPSEstimateCurrentYear', 'PriceEPSEstimateNextYear',
                'PriceSales', 'ShortRatio', 'StockExchange', 'Symbol',
                'TwoHundreddayMovingAverage', 'Volume', 'YearHigh',
                'YearLow', 'YearRange']

        response = request_quotes([self.__ticker], keys)
        # if not response['Name']:
        #     raise RequestError(
        #         self.__ticker + ' returns no results from Yahoo Finance.'
        #     )
        return response

    def get_historical(self, start_date, end_date):
        """Get stock's daily historical information.

        Returns a dictionary with Adj Close, Close, High, Low, Open and
        Volume, between the start_date and the end_date. Is start_date and
        end_date were not provided all the available information will be
        retrieved. Information provided by YQL platform.
        Check `here <http://goo.gl/8AROUD>`_ for more information on YQL.

        .. warning:: Request limited to a period not greater than 366 days.
            Use download_historical() to download the full historical data.


        >>> stock.get_historical('2016-03-01', '2016-03-02')
        [
            {
                'Close': '100.75',
                'Low': '99.639999',
                'High': '100.889999',
                'Adj_Close': '100.140301',
                'Date': '2016-03-02',
                'Open': '100.510002',
                'Volume': '33169600'
            },
            {
                'Close': '100.529999',
                'Low': '97.419998',
                'High': '100.769997',
                'Adj_Close': '99.921631',
                'Date': '2016-03-01',
                'Open': '97.650002',
                'Volume': '50407100'
            }
        ]

        :param start_date: Start date
        :type start_date: string on the format of "yyyy-mm-dd"
        :param end_date: End date
        :type end_date: string on the format of "yyyy-mm-dd"
        :returns: Daily historical information.
        :rtype: list of dictionaries
        """
        return request_historical(self.__ticker, start_date, end_date)

    def save_historical(self, output_folder):
        """Download historical data from Yahoo Finance.

        Downloads full historical data from Yahoo Finance as CSV. The following
        fields are available: Adj Close, Close, High, Low, Open and Volume.
        Files will be saved to output_folder as <ticker>.csv.

        :param output_folder: Output folder path
        :type output_folder: string
        """
        download_historical([self.__ticker], output_folder)
